File: solutions/python/yacht/1/yacht.py
from collections import Counter

# Score categories.
# Change the values as you see fit.
YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11

YACHT_LIST = [ONES, TWOS, THREES, FOURS, FIVES, SIXES]

def score(dice, category):
    dice_counter = Counter(dice)
    if category == YACHT:
        return 50 if any(v == 5 for v in dict(dice_counter).values()) else 0
    if category in YACHT_LIST:
        cat_index = YACHT_LIST.index(category) + 1
        return cat_index* dice_counter[cat_index]
    if category == FULL_HOUSE:
        return 0 if set(v for v in dict(dice_counter).values()) ^ {2,3} else sum(dice)
    if category == LITTLE_STRAIGHT:
        return 30 if sorted(dice) == [1,2,3,4,5] else 0
    if category == BIG_STRAIGHT:
        return 30 if sorted(dice) == [2,3,4,5,6] else 0
    if category == CHOICE:
        return sum(dice)
    if category == FOUR_OF_A_KIND:
        return sum( 4 * k for k,v in dict(dice_counter).items() if v >= 4)
    raise ValueError("Invalid Category")

File: solutions/java/yacht/1/src/main/java/Yacht.java
import java.util.List;
import java.util.stream.IntStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;


class Yacht {

    private int score;
    private static List <YachtCategory> yachtCategories;
    
    static{
        yachtCategories = Arrays.asList(YachtCategory.ONES, YachtCategory.TWOS, 
            YachtCategory.THREES, YachtCategory.FOURS,
            YachtCategory.FIVES, YachtCategory.SIXES);
    }

    Yacht(int[] dice, YachtCategory yachtCategory) {
        
        int numDice = dice.length;
        if (numDice != 5) throw new IllegalArgumentException();

        List<Integer> histogram = new ArrayList<Integer>(Collections.nCopies(6, 0));
        int sumDices = 0;
        for (int i = 0; i < numDice; ++i) {
            sumDices += dice[i];
            int d = dice[i] - 1;
            histogram.set(d, histogram.get(d) + 1);
        }

        if (yachtCategory == YachtCategory.YACHT) 
            this.score = histogram.stream().anyMatch(x -> x == 5) ? 50 : 0;

        int yachtCategoryIndex = yachtCategories.indexOf(yachtCategory);
        if (yachtCategoryIndex >= 0)
            this.score = (yachtCategoryIndex + 1) * histogram.get(yachtCategoryIndex);

        if (yachtCategory == YachtCategory.FULL_HOUSE) {
            boolean isFullHouse = histogram.indexOf(3) >= 0 && histogram.indexOf(2) >= 0;
            this.score = isFullHouse ? sumDices : 0;
        }

        if (yachtCategory == YachtCategory.LITTLE_STRAIGHT)
            this.score = histogram.subList(0, 5).stream().allMatch(x -> x == 1) ? 30 : 0;

        if (yachtCategory == YachtCategory.BIG_STRAIGHT)
            this.score = histogram.subList(1, 6).stream().allMatch(x -> x == 1) ? 30 : 0;

        if (yachtCategory == YachtCategory.CHOICE)
            this.score = sumDices;

        if (yachtCategory == YachtCategory.FOUR_OF_A_KIND)
            this.score = IntStream.range(0, 6).filter(x -> histogram.get(x) > 3).map(x -> x + 1).sum() * 4;

    }

    int score() {
        return this.score;
    }

}
